Add BLAS level 1 routine snrm2 (#286)
stdlib-js/stdlib: lib/node_modules/@stdlib/blas/base/snrm2/README.md

var Float32Array = require( '@stdlib/array/float32' );
var floor = require( '@stdlib/math/base/special/floor' );

var x0 = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
var x1 = new Float32Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

var N = floor( x0.length / 2 );

var z = snrm2( N, x1, 2 );
// returns 5.0
```

#### snrm2.ndarray( N, x, stride, offset )

Computes the [L2-norm][l2-norm] of the given array with alternative indexing semantics.

```javascript
var Float32Array = require( '@stdlib/array/float32' );

var x = new Float32Array( [ 1.0, -2.0, 2.0 ] );
var N = 3;

var z = snrm2.ndarray( N, x, 1, 0 );
// returns 3.0
```

The function accepts the following additional parameters:

-   **offset**: starting index for `x`.

While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on starting indicies.  For example, to calculate the [L2-norm][l2-norm] for every other value in `x` starting from the second value

```javascript
var Float32Array = require( '@stdlib/array/float32' );
var floor = require( '@stdlib/math/base/special/floor' );

> var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);

Reviewer: ```suggestion
var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
```

<!--

@license Apache-2.0

Copyright (c) 2020 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# snrm2

> Calculate the L2-norm of a vector `x`.

<section class="intro>">

> The [L2 norm][l2norm] is defined as

Reviewer: ```suggestion
The [L2-norm][l2-norm] is defined as
```

> <!-- <equation class="equation" label="eq:l2norm" align="center raw="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" -->

Reviewer: ```suggestion
<!-- <equation class="equation" label="eq:l2_norm" align="center raw="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" -->
```

> <div class="equation" align="center" data-raw-text="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" data-equation="eq:l2norm">

Reviewer: ```suggestion
<div class="equation" align="center" data-raw-text="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" data-equation="eq:l2_norm">
```

>     <img src="" alt="L2 norm definition.">

Reviewer: ```suggestion
    <img src="" alt="L2-norm definition.">
```

    <br>
</div>

<!-- </equation> -->

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var snrm2 = require( '@stdlib/blas/base/snrm2' );
```

#### snrm2( N, x, stride )

> Computes the L2-norm (length) of the given array.

Reviewer: ```suggestion
Computes the [L2-norm][l2-norm] of the given array.
```

```javascript
var Float32Array = require( '@stdlib/array/float32' );

var x = new Float32Array( [ 1.0, -2.0, 2.0 ] );
var N = 3;

var z = snrm2( N, x, 1 );
// returns 3.0
```

The function accepts the following parameters:

-   **N**: number of elements in the array.

> -   **x**: input [`Float32Array`][@stdlib/array/float32]

Reviewer: ```suggestion
-   **x**: input [`Float32Array`][@stdlib/array/float32].
```

-   **stride**: index increment for `x`.

> The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For example, to compute the length of every other element of `x`,

Reviewer: ```suggestion
The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For example, to compute the [L2-norm][l2-norm] of every other element of `x`,
```

```javascript
var Float32Array = require( '@stdlib/array/float32' );
var floor = require( '@stdlib/math/base/special/floor' );

var x = new Float32Array( [ 1.0, 2.0, 2.0, -7.0, -2.0, 3.0, 4.0, 2.0 ] );
var N = floor( x.length / 2 );

var z = snrm2( N, x, 2 );
// returns 5.0
```

Note that indexing is relative to the first index.  To introduce an offset, use [`typed array`][mdn-typed-array] views.

```javascript
var Float32Array = require( '@stdlib/array/float32' );
var floor = require( '@stdlib/math/base/special/floor' );

var x0 = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
var x1 = new Float32Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

var N = floor( x0.length / 2 );

var z = snrm2( N, x1, 2 );
// returns 5.0
```

#### snrm2.ndarray( N, x, stride, offset )

> Computes the L2-norm (length) of the given array with alternative indexing semantics.

Reviewer: ```suggestion
Computes the [L2-norm][l2-norm] of the given array with alternative indexing semantics.
```

```javascript
var Float32Array = require( '@stdlib/array/float32' );

var x = new Float32Array( [ 1.0, -2.0, 2.0 ] );
var N = 3;

var z = snrm2.ndarray( N, x, 1, 0 );
// returns 3.0
```

The function accepts the following additional parameters:

-   **offset**: starting index for `x`.

> While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on starting indicies.  For example, to calculate the L2-norm for every other value in `x` starting from the second value

Reviewer: ```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on starting indicies.  For example, to calculate the [L2-norm][l2-norm] for every other value in `x` starting from the second value
```

```javascript
var Float32Array = require( '@stdlib/array/float32' );
var floor = require( '@stdlib/math/base/special/floor' );

var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);

> 

Reviewer: ```suggestion
```

var N = floor( x.length / 2 );

var z = snrm2.ndarray( N, x, 2, 1 );
// returns 5.0
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   If `N <= 0` or `stride <= 0`, both functions return `0.0`.
-   `snrm2()` corresponds to the [BLAS][blas] level 1 function [`snrm2`][snrm2].

</section>

<!-- /.notes -->

<section class="examples">

## Examples

```javascript
var randu = require( '@stdlib/random/base/randu' );
var round = require( '@stdlib/math/base/special/round' );
var Float32Array = require( '@stdlib/array/float32' );
var snrm2 = require( '@stdlib/blas/base/snrm2' );

var x;
var i;

x = new Float32Array( 10 );
for ( i = 0; i < x.length; i++ ) {
    x[ i ] = round( randu() * 100.0 );
}
console.log( x );

var z = snrm2( x.length, x, 1 );
console.log(z);
```

</section>

<!-- /.examples -->

<section class="links">

> [l2norm]: https://en.wikipedia.org/wiki/Euclidean_distance

Reviewer: ```suggestion
[l2-norm]: https://en.wikipedia.org/wiki/Euclidean_distance
```

<!--

@license Apache-2.0

> Copyright (c) 2019 The Stdlib Authors.

Reviewer: Would you mind updating all the copyright years in this PR to `2020`? Thanks!

Author: No Problem

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# snrm2

> Calculate the 2-norm of a vector `x`.

<section class="intro>">

The [L2 norm][l2norm] is defined as

<!-- <equation class="equation" label="eq:l2norm" align="center raw="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" -->

<div class="equation" align="center" data-raw-text="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" data-equation="eq:l2norm">

>     ![L2 norm equation][equation]

Reviewer: Ah. We do this a bit differently. See `ddot` for an example. Leave the `src` attribute as an empty string, and I'll update it on my end after merging this PR.

    <br>
</div>

<!-- </equation> -->

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var snrm2 = require( '@stdlib/blas/base/snrm2' );
```

#### snrm2( N, x, stride )

Computes the 2-norm (length) of the given array.

```javascript
var Float32Array = require( '@stdlib/array/float32' );

var x = new Float32Array( [ 1.0, -2.0, 2.0 ] );
var N = 3;

var z = snrm2( N, x, 1 );
// returns 3.0
```

The function accepts the following parameters:

-   **N**: number of elements in the array.
-   **x**: input [`Float32Array`][@stdlib/array/float32]
-   **stride**: index increment for `x`.

The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For

> example, to compute the length of every other element of `x`,

Reviewer: You can remove the linebreak here.

> example, to compute the length of every other element of `x`,

Reviewer: ```suggestion
example, to compute the 2-norm for every other element of `x`,
```

```javascript
var Float32Array = require( '@stdlib/array/float32' );
var floor = require( '@stdlib/math/base/special/floor' );

> var x = new Float32Array( [ 1.0, 2.0, 2.0, -7.0, -2.0, 3.0, 4.0, 2.0 ]);

Reviewer: ```suggestion
var x = new Float32Array( [ 1.0, 2.0, 2.0, -7.0, -2.0, 3.0, 4.0, 2.0 ] );
```

> var N = floor(x.length / 2);

Reviewer: ```suggestion
var N = floor( x.length / 2 );
```

var z = snrm2( N, x, 2 );
// returns 5.0
```

> Note that indexing is relative to the first index.  To introduce and offset, use [`typed array`][mdn-typed-array] views.

Reviewer: ```suggestion
Note that indexing is relative to the first index.  To introduce an offset, use [`typed array`][mdn-typed-array] views.
```

```javascript
var Float32Array = require( '@stdlib/array/float32' );
var floor = require( '@stdlib/math/base/special/floor' );

var x0 = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
var x1 = new Float32Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

> var N = floor(x.length / 2);

Reviewer: ```suggestion
var N = floor( x.length / 2 );
```

var z = snrm2( N, x1, 2 );
// returns 5.0
```

#### snrm2.ndarray( N, x, stride, offset )

Computes the 2-norm (length) of the given array with alternative indexing semantics.

```javascript
var Float32Array = require( '@stdlib/array/float32' );

var x = new Float32Array( [ 1.0, -2.0, 2.0 ] );
var N = 3;

var z = snrm2.ndarray( N, x, 1, 0 );
// returns 3.0
```

The function accepts the following additional parameters:

-   **offset**: starting index for `x`.

> While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on starting indicies.  For example to calculate the 2-norm pf every other value in `x` starting from the second value

Reviewer: ```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on starting indicies.  For example, to calculate the 2-norm for every other value in `x` starting from the second value
```

```javascript
var Float32Array = require( '@stdlib/array/float32' );
var floor = require( '@stdlib/math/base/special/floor' );

var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);

> var N = floor(x.length / 2);

Reviewer: ```suggestion
var N = floor( x.length / 2 );
```

var z = snrm2.ndarray( N, x, 2, 1 );
// returns 5.0
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

> -   If `N <= 0` both functions return `0.0`.

Reviewer: ```suggestion
-   If `N <= 0` or `stride <= 0`, both functions return `0.0`.
```

> -   If `stride <= 0` both functions return `0.0`.

Reviewer: ```suggestion
```

-   `snrm2()` corresponds to the [BLAS][blas] level 1 function [`snrm2`][snrm2].

</section>

<!-- /.notes -->

<section class="examples">

## Examples

```javascript
var randu = require( '@stdlib/random/base/randu' );
var round = require( '@stdlib/math/base/special/round' );
var Float32Array = require( '@stdlib/array/float32' );
var snrm = require( '@stdlib/blas/base/snrm2' );

var x;
var i;

x = new Float32Array( 10 );
for ( i = 0; i < x.length; i++ ) {
    x[ i ] = round( randu() * 100.0 );
}
console.log( x );

var z = snrm2( x.length, x, 1 )l
console.log(z);
```

</section>

<!-- /.examples -->

<section class="links">

[l2norm]: https://en.wikipedia.org/wiki/Euclidean_distance

[blas]: http://www.netlib.org/blas

[snrm2]: http://www.netlib.org/lapack/explore-html/df/d28/group__single__blas__level1.html

> [equation]: https://latex.codecogs.com/svg.latex?%5Cinline%20%5C%7C%5Cmathbf%7Bx%7D%5C%7C_2%20%3D%20%5Csqrt%7B%5Cmathbf%7Bx%7D%5ET%5Cmathbf%7Bx%7D%7D

Reviewer: ```suggestion
```
